substrate/network-libp2p/src/service.rs

```diff
@@ -978,6 +1000,58 @@ fn connect_to_nodes<T, To, St, C>(
 	}
 }
 
+
+fn connect_with_query_peer_id<T, To, St, C>(
+	shared: Arc<Shared>,
+	base_transport: T,
+	proto: RegisteredProtocol<Arc<NetworkProtocolHandler + Send + Sync>>,
+	addr: Multiaddr,
+	swarm_controller: &SwarmController<St>
+)
+	where T: MuxedTransport<Output =  TransportOutput<To>> + Clone + 'static,
+		T::MultiaddrFuture: 'static,
+		To: AsyncRead + AsyncWrite + 'static,
+		St: MuxedTransport<Output = FinalUpgrade<C>> + Clone + 'static,
+		C: 'static,
+{
+	let addr2 = addr.clone();
+	let with_proto = base_transport
+		.clone()
+		.and_then(move |out, endpoint, client_addr| {
+			trace!(target: "sub-libp2p", "in");
+			let socket = out.socket;
+			let original_addr = out.original_addr;
+			out.info
+				.and_then(move |info| {
+					let _ = process_identify_info(shared, &info, original_addr,
+						endpoint, &base_transport);
+					trace!(target: "sub-libp2p", "Bootnode {:} found with peer id: {:?}",
+						addr2, info.info.public_key.into_peer_id());
+					upgrade::apply(socket, proto, endpoint, client_addr)
+				})
+		})
+		.and_then(move |out, _endpoint, client_addr|
```

Comment:
Down there the code is duplicated, which is not great because from a logic point of view we always want the behaviour to be the same.

Reply:
while I agree that there should be one way of connecting to a node and not two, they are distinct in that the other one _requires_ the peerID and does a lot more things indirectly based on that (like not calling the swarm-controller directly), rather than the direct way in which this part works. Combining the two, however, I don't see possible unless we refactor major parts of two crates to not expect a peerID in the first place. While I am totally in favor of that, I wouldn't do that as part of this change.